packages/rest/src/actions/middleware-action.ts

```diff
@@ -0,0 +1,84 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  config,
+  Constructor,
+  InvocationResult,
+  Next,
+  Provider,
+} from '@loopback/context';
+import {RequestHandler} from 'express';
+import {HttpContext} from '../types';
+import {BaseRestAction} from './base-action';
+
+/**
+ * Wrap an Express middleware handler as RestAction
+ */
+export class MiddlewareAction extends BaseRestAction {
+  constructor(private requestHandler: RequestHandler) {
+    super();
+  }
+
+  async intercept(ctx: HttpContext, next: Next) {
+    return new Promise<InvocationResult>((resolve, reject) => {
+      try {
+        if (this.requestHandler.length < 3) {
+          // The express middleware does not call `next`
+          this.requestHandler(ctx.request, ctx.response, () => {});
+          resolve(next());
```

Comment:
This is problematic. Consider the following example.

```ts
// middleware
function myRoute(req, res) {
  doSomeAsyncWork((err, result) => {
    console.log('async work is done');
    res.json(result);
  });
}

// interceptor/sequence action
async function logRequest(ctx, next) {
  console.log(
    'START %s %s',
    ctx.req.method, 
    ctx.req.path, 
  );

  await next();

  console.log(
    'DONE %s %s %s',
    ctx.req.method, 
    ctx.req.path, 
    ctx.res.statusCode, 
  );
}
```

This will produce the following console logs:

```
START GET /api/users
DONE GET /api/users
async work is done
```

Expected output:

```
START GET /api/users
async work is done
DONE GET /api/users
```

IMO, when the middleware does not call `next`, you should use `on-finished` module to wait for the response to be written before you call `next`.

See the existing implementation here:

https://github.com/strongloop/loopback-next/blob/0b5a47227ac0ab71bf67cd8836dfe3ecbbe9a17d/packages/rest/src/router/external-express-routes.ts#L160-L177